clamscan/manager.c

```diff
@@ -631,33 +631,325 @@ static int scanstdin(const struct cl_engine *engine, struct cl_scan_options *opt
+ * @brief Progress callback for sig-load
+ *
+ * @param total_items   Total number of items
+ * @param now_completed Number of items completed
+ * @param context       Opaque application provided data; This maps to engine_compile_progress
+ */
+static void engine_compile_callback(size_t total_items, size_t now_completed, void *context)
+{
+    time_t curtime = 0;
+    time_t remtime = 0;
+
+    struct engine_compile_progress *engineCompileProgress = (struct engine_compile_progress *)context;
+
+    uint32_t i                  = 0;
+    uint32_t totalNumDots       = 25;
+    uint32_t numDots            = 0;
+    double fractiondowncompiled = 0.0;
+
+    if ((total_items <= 0) || (engineCompileProgress->bComplete)) {
+        return;
+    }
+
+    fractiondowncompiled = (double)now_completed / (double)total_items;
+    numDots              = round(fractiondowncompiled * totalNumDots);
+
+    if (0 == engineCompileProgress->startTime) {
+        engineCompileProgress->startTime = time(0);
+    }
+    curtime = time(0) - engineCompileProgress->startTime;
+
+    engineCompileProgress->lastRunTime = curtime;
+
+#ifndef _WIN32
+    fprintf(stdout, "\e[?7l");
+#endif
+    if (fractiondowncompiled <= 0.0) {
+        fprintf(stdout, "Compiling: ");
+        print_time(curtime);
+        fprintf(stdout, "               ");
+    } else {
+        remtime = (curtime / fractiondowncompiled) - curtime;
+        fprintf(stdout, "Compiling: ");
+        print_time(curtime);
+        fprintf(stdout, ", ETA: ");
+        print_time(remtime);
+        fprintf(stdout, " ");
+    }
+
+    fprintf(stdout, "[");
+    if (numDots > 0) {
+        if (numDots > 1) {
+            for (i = 0; i < numDots - 1; i++) {
+                fprintf(stdout, "=");
+            }
+        }
+        fprintf(stdout, ">");
+        i++;
+    }
+    for (; i < totalNumDots; i++) {
+        fprintf(stdout, " ");
+    }
+
+    fprintf(stdout, "] ");
+
+    print_no_sigs(now_completed, 1);
+    fprintf(stdout, "/");
+    print_no_sigs(total_items, 0);
+    fprintf(stdout, " tasks ");
+
+    if (now_completed < total_items) {
+        fprintf(stdout, "\r");
+    } else {
+        fprintf(stdout, "\n");
+        engineCompileProgress->bComplete = 1;
+    }
+#ifndef _WIN32
+    fprintf(stdout, "\e[?7h");
+#endif
+    fflush(stdout);
+
+    return;
+}
+
+#ifdef ENABLE_ENGINE_FREE_PROGRESSBAR
+/**
+ * @brief Progress callback for sig-load
+ *
+ * @param total_items   Total number of items
+ * @param now_completed Number of items completed
+ * @param context       Opaque application provided data; This maps to engine_free_progress
+ */
+static void engine_free_callback(size_t total_items, size_t now_completed, void *context)
+{
+    time_t curtime = 0;
+
+    struct engine_free_progress *engineFreeProgress = (struct engine_free_progress *)context;
+
+    uint32_t i                = 0;
+    uint32_t totalNumDots     = 10;
+    uint32_t numDots          = 0;
+    double fractiondownloaded = 0.0;
```

Comment:
`fractiondownunloaded`? (Maybe I misunderstand the `down` part of this though)

Reply:
That's a copy-paste typo.  I copied the implementation from the freshclam progress bar, which had used a curl callback to track download and upload progress (not that we used the upload one).  I should rename this variable.